### ISSUE TITLE
exmod/ex03.sh fails on Linux

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/gmt#940]
**Description of the problem**

This is what I get on Linux:

![image](https://user-images.githubusercontent.com/3974108/59555839-3a1f5680-8f87-11e9-843d-b44de616b1de.png)

The differences between PS files generated on Linux and mac is attached. It seems there is an issue with subplot frame settings.

```diff
diff ex03-linux.ps ex03-mac.ps
8c8
< %%CreationDate: Sat Jun 15 15:57:17 2019
---
> %%CreationDate: Sat Jun 15 16:03:27 2019
688c688
< %@GMT: gmt psxy spectrum.xpower -JX-4il/3.75il -Bxa1f3p+l/255/240 '-Bya1f3p+lPower (mGal@+2@+km)' -Gred -ST0.07i -R1/1000/0.1/10000 -Ey+p0.5p -BWens+g24 -Xa0i -Ya4.0194i
---
> %@GMT: gmt psxy spectrum.xpower -JX-4il/3.75il -Bxa1f3p '-Bya1f3p+lPower (mGal@+2@+km)' -Gred -ST0.07i -R1/1000/0.1/10000 -Ey+p0.5p -BWens+g240/255/240 -Xa0i -Ya4.0194i
702c702
< {0.0941 A} FS
---
> {0.941 1 0.941 C} FS
1904c1904
< %@GMT: gmt psxy spectrum.coh -JX-4il/3.75i -Bxa1f3p+l/255/240 -Bya0.25f0.05+lCoherency@+2@+ -R1/1000/0/1 -Sc0.07i -Gpurple -Ey+p0.5p -BWenS+g24 -Xa0i -Ya-0.0694i
---
> %@GMT: gmt psxy spectrum.coh -JX-4il/3.75i '-Bxa1f3p+lWavelength (km)' -Bya0.25f0.05+lCoherency@+2@+ -R1/1000/0/1 -Sc0.07i -Gpurple -Ey+p0.5p -BWenS+g240/255/240 -Xa0i -Ya-0.0694i
1920c1920
< {0.0941 A} FS
---
> {0.941 1 0.941 C} FS
2853c2853
< (/255/240) bc Z
---
> (Wavelength \(km\)) bc Z
```

**Full script that generated the error**

```
PASTE CODE HERE
```

**Comment by PaulWessel** [comment]
Hm, not clear, but part of the **+g**<_r/g/b_> ends up as part of a **+l** label argument to **-S** in subplot? Unfortunately, did not bring home Linux laptop but am going in later so can bring it and do some good ole' debugging on Fathers Day. If you are able to experiment a bit on Linux, see if it makes any difference if you:

1. Give a named color via **+g**, and try a short name (pink) versus long (LightSteelBlue1)
2. Remove the **+l**"Wavelength (km)" from subplot and see if still failing.  Remove the (km) only to see if parenthesis are a problem.
3. Move the **-B** option ahead of **-S** in the subplot call.

This could provide useful hints for the debug.

**Comment by seisman** [comment]
> Give a named color via +g, and try a short name (pink) versus long (LightSteelBlue1)

Changing `-BWeSn+g240/255/240` to `-BWeSn+gpink` results in errors:
```
psxy [ERROR]: Colorname pi not recognized!
psxy [ERROR]: Bad +g<fill> argument pi
psxy [ERROR]: Syntax error -B option. Correct syntax:
	-B[p|s][x|y|z]<intervals>[+a<angle>|n|p][+l|L<label>][+p<prefix>][+u<unit>] -B[<axes>][+b][+g<fill>][+o<lon>/<lat>][+t<title>] OR
	-B[p|s][x|y|z][a|f|g]<tick>[m][l|p] -B[p|s][x|y|z][+l<label>][+p<prefix>][+u<unit>] -B[<axes>][+b][+g<fill>][+o<lon>/<lat>][+t<title>]
psxy [ERROR]: Offending option -BWens+gpi
```

Using long colorname `LightSteelBlue1`:
```
psxy [ERROR]: Colorname Li not recognized!
psxy [ERROR]: Bad +g<fill> argument Li
psxy [ERROR]: Syntax error -B option. Correct syntax:
	-B[p|s][x|y|z]<intervals>[+a<angle>|n|p][+l|L<label>][+p<prefix>][+u<unit>] -B[<axes>][+b][+g<fill>][+o<lon>/<lat>][+t<title>] OR
	-B[p|s][x|y|z][a|f|g]<tick>[m][l|p] -B[p|s][x|y|z][+l<label>][+p<prefix>][+u<unit>] -B[<axes>][+b][+g<fill>][+o<lon>/<lat>][+t<title>]
psxy [ERROR]: Offending option -BWens+gLi
```

> Remove the +l"Wavelength (km)" from subplot and see if still failing. Remove the (km) only to see if parenthesis are a problem.

Diff shows no changes. It seems `+l"Wavelength (km)"` is never used.

> Move the -B option ahead of -S in the subplot call.

Nothing changes.

**Comment by PaulWessel** [comment]
Interesting, so the -BWeSn+gpink is truncated to -BWeSn+gpi, but I do not (yet) see any such limitation in the code (and clearly this does not happen in macOS).  Will need to debug in ddd I'm afraid.  it is not in the **-B** parsing but music be something subtle in subplot that sets up those parameters.

**Comment by PaulWessel** [comment]
Could you run the script manually, i.e., running one command at the time, and after running the subplot command, find the session directory in ~/.gmt/sessions and send the gmt subplot parameter file created (has the name subplot in it). That file has the args used to create the options to the panel calls.

**Comment by seisman** [comment]
I added a `sleep 30` command and checked the session dir. 

This is what I got.
```
$ cat gmt.subplot.0
# subplot panel information file
# Command: subplot begin 2x1 -M0.1i -BWeSn+g240/255/240 -S"Cb+lWavelength (km)" -T"Ship and Satellite Gravity" -Fs4i/3.75i -A+jTR+o0.1i -X2i -Y1.5i
# HEADING: 2 8.11944 Ship and Satellite Gravity
# ORIGIN: 0 0
# DIMENSION: 4 7.76944
# PARALLEL: 0
#panel	row	col	nrow	ncol	x0	y0	w	h	tag	tag_dx	tag_dy	tag_clearx	tag_cleary	tag_pos	tag_just	tag_fill	tag_pen	Bframe	Bx	By	Ax	Ay
0	0	0	2	1	0.0000	4.0194	4.0000	3.7500	a)	0.1	0.1	0.0375	0.0375	TR	TR	-	-	Wens+g240/255/240afaf
1	1	0	2	1	0.0000	-0.0694	4.0000	3.7500	b)	0.1	0.1	0.0375	0.0375	TR	TR	-	-	WenS+g240/255/240Wavelength (km)afaf

$ cat gmt.subplotorder.0
2 1 0
```

**Comment by PaulWessel** [comment]
Could you attach it as a file? It has non printable characters between fields so hard to tell.

**Comment by seisman** [comment]
Attached:
[gmt.subplotorder.0.txt](https://github.com/GenericMappingTools/gmt/files/3293589/gmt.subplotorder.0.txt)
[gmt.subplot.0.txt](https://github.com/GenericMappingTools/gmt/files/3293590/gmt.subplot.0.txt)

**Comment by PaulWessel** [comment]
Thanks, those look fine to me, so I will need to look at where they are read back in and used.

**Comment by PaulWessel** [comment]
Think I found it - I must have forgotten about any +modifiers to -Bwesn etc in this context.  Could you edit this line in gmt_types.h from

`	char Baxes[GMT_LEN8];		/* The -B setting for selected axes */
`

to

`	char Baxes[GMT_LEN64];		/* The -B setting for selected axes */
`

and see if it fixes it?

**Comment by seisman** [comment]
Yes, it works. But why does the old one work on mac?

**Comment by PaulWessel** [comment]
Because macOS is helping behind the scenes and somehow overwriting the address beyond 7 is not causing any errors.  Perhaps the padding is much larger.  We need to play with thins like https://sourceforge.net/projects/boundschecking/, but not sure if this is it.  In Xcode there are many settings but most for Malloc stuff.  This is straight off the heap and only something like the bounds checking will catch it.
Will you commit the fix or shall I?

**Comment by PaulWessel** [comment]
@joa-quim, it would be great if you could build the latest and run the full tests and post the build/Testing/Temporary/LastTest.log afterwards so we can look for any tests that fail in odd ways, pointing to bugs that show up on your end but are not harmful on macOS.

**Comment by seisman** [comment]
You can do it.

**Comment by PaulWessel** [comment]
valgrind would find this problem and works on Linux.

**Comment by joa-quim** [comment]
Right. This afternoon I managed to stick a 2 cm piece of wood in my foot during my agriculture activities. 
But feet are resistant pieces of our body and I can walk better than I thought.

Testing on Win is difficult. MingW bash makes some crazy moves. This are the failures I get now (despite telling it not to, it insists in running the movie tests)
```
The following tests FAILED:
        150 - anim01/anim_01.sh (Failed)
        151 - anim02/anim_02.sh (Failed)
        152 - anim03/anim_03.sh (Failed)
        153 - anim04/anim_04.sh (Failed)
        154 - anim05/anim_05.sh (Failed)
        155 - anim06/anim_06.sh (Failed)
        156 - anim07/anim_07.sh (Failed)
        157 - anim08/anim_08.sh (Failed)
        160 - ex03/example_03.sh (Failed)
        196 - ex39/example_39.sh (Failed)
        198 - ex41/example_41.sh (Failed)
        201 - ex44/example_44.sh (Failed)
        224 - exmod/ex03.sh (Failed)
        237 - filter1d/highpass.sh (Failed)
        240 - fitcircle/circles.sh (Failed)
        261 - geodesy/case_largeR_withW.sh (Failed)
        269 - geodesy/gpsgridder1.sh (Failed)
        270 - geodesy/gpsgridder2.sh (Failed)
        280 - gmtconvert/segsearch.sh (Failed)
        306 - gmtspatial/poldecimate.sh (Failed)
        381 - grdimage/globalgrid.sh (Failed)
        434 - grdvector/vectors.sh (Failed)
        451 - greenspline/gspline_5.sh (Failed)
        462 - mapproject/oblmerc_down.sh (Failed)
        464 - mapproject/proj4.sh (Failed)
        470 - mgd77/test_cm4.sh (Failed)
        489 - potential/admittance.sh (Failed)
        510 - project/cartesian.sh (Failed)
        524 - psbasemap/map_JE.sh (Failed)
        526 - psbasemap/mapscales.sh (Failed)
        539 - psbasemap/scales.sh (Failed)
        559 - pscoast/pseudo.sh (Failed)
        624 - pssolar/shading.sh (Failed)
        645 - psxy/antwrap.sh (Failed)
        658 - psxy/categorical.sh (Failed)
        674 - psxy/curvedmap_caps.sh (Failed)
        676 - psxy/custom_symbol.sh (Failed)
        684 - psxy/gallo.sh (Failed)
        712 - psxy/plane_cart.sh (Failed)
        713 - psxy/plane_geo.sh (Failed)
        797 - spotter/spotter_06.sh (Failed)
        823 - x2sys/x2sys_04.sh (Failed)
        824 - x2sys/x2sys_05.sh (Failed)
```
Oddly, I see tons of lines in the ``LatestTest.log`` with
```
The following tests FAILED:
        150 - anim01/anim_01.sh (Failed)
        151 - anim02/anim_02.sh (Failed)
        152 - anim03/anim_03.sh (Failed)
        153 - anim04/anim_04.sh (Failed)
        154 - anim05/anim_05.sh (Failed)
        155 - anim06/anim_06.sh (Failed)
        156 - anim07/anim_07.sh (Failed)
        157 - anim08/anim_08.sh (Failed)
        160 - ex03/example_03.sh (Failed)
        196 - ex39/example_39.sh (Failed)
        198 - ex41/example_41.sh (Failed)
        201 - ex44/example_44.sh (Failed)
        224 - exmod/ex03.sh (Failed)
        237 - filter1d/highpass.sh (Failed)
        240 - fitcircle/circles.sh (Failed)
        261 - geodesy/case_largeR_withW.sh (Failed)
        269 - geodesy/gpsgridder1.sh (Failed)
        270 - geodesy/gpsgridder2.sh (Failed)
        280 - gmtconvert/segsearch.sh (Failed)
        306 - gmtspatial/poldecimate.sh (Failed)
        381 - grdimage/globalgrid.sh (Failed)
        434 - grdvector/vectors.sh (Failed)
        451 - greenspline/gspline_5.sh (Failed)
        462 - mapproject/oblmerc_down.sh (Failed)
        464 - mapproject/proj4.sh (Failed)
        470 - mgd77/test_cm4.sh (Failed)
        489 - potential/admittance.sh (Failed)
        510 - project/cartesian.sh (Failed)
        524 - psbasemap/map_JE.sh (Failed)
        526 - psbasemap/mapscales.sh (Failed)
        539 - psbasemap/scales.sh (Failed)
        559 - pscoast/pseudo.sh (Failed)
        624 - pssolar/shading.sh (Failed)
        645 - psxy/antwrap.sh (Failed)
        658 - psxy/categorical.sh (Failed)
        674 - psxy/curvedmap_caps.sh (Failed)
        676 - psxy/custom_symbol.sh (Failed)
        684 - psxy/gallo.sh (Failed)
        712 - psxy/plane_cart.sh (Failed)
        713 - psxy/plane_geo.sh (Failed)
        797 - spotter/spotter_06.sh (Failed)
        823 - x2sys/x2sys_04.sh (Failed)
        824 - x2sys/x2sys_05.sh (Failed)
```
Oddly, I see lots of lines in my ``LatestTest.log`` with
```
gmt.exe [ERROR]: Unable to open GMT shared mbgmt library: Error <193>:
```
but that lib (MB-system) for me is ``mbgmt_w64.dll``. Anyway, the failures above are mostly known old ones. Some (gmtspatial/poldecimate.sh) happen in Linux too.

**Comment by joa-quim** [comment]
Ah, and somehow when running the tests sometimes it thinks it's running on modern and fails with
```
489/826 Testing: potential/admittance.sh
489/826 Test: potential/admittance.sh
Command: "C:/MinGW64/msys64/usr/bin/bash.exe" "gmtest" "potential/admittance.sh"
Directory: C:/v/build14/test
"potential/admittance.sh" start time: Jun 15 20:12 GMT Daylight Time
Output:
----------------------------------------------------------
gmt.exe [ERROR]: Option -O not allowed for modern GMT mode.
gmt.exe [ERROR]: Option -K not allowed for modern GMT mode.
gmt.exe [ERROR]: Shorthand -J not allowed for modern GMT mode.
ERROR: gmtest:58
```

**Comment by PaulWessel** [comment]
I get that (infrequently) too.  I think it happens when a modern mode script crashes and leaves stuff, and with your OS lack of PPID aren't they all running with same PPID?  Maybe we need a scheme to increase your fake PPID by one for each test?

**Comment by PaulWessel** [comment]
Wooden stakes through feet are unnecessary, just rub with garlic to keep vampires away.
Could you post the large log though?

**Comment by joa-quim** [comment]
I fixed PPID issue for the *nix pretenders (batch always worked fine) has been fixed a couple months ago.

**Comment by joa-quim** [comment]
OK, but it's very frustrating. Almost always it turns out to be an issue in the script because f some MinGW bash shit.

[LastTest.log](https://github.com/GenericMappingTools/gmt/files/3293635/LastTest.log)

**Comment by joa-quim** [comment]
> Wooden stakes through feet are unnecessary
Specially of that size.

**Comment by PaulWessel** [comment]
Some stuff is intriguing:  Here you/we have a double C:C:

psxy [ERROR]: Could not find either custom symbol or EPS macro C:C:/MinGW64/msys64/progs_cygw/GMTdev/gmt5/master/test/psxy/plane_c

"x2sys/x2sys_04.sh" start time: Jun 15 20:17 GMT Daylight Time
Output:
----------------------------------------------------------
gmt.exe [ERROR]: Unable to open GMT shared mbgmt library: Error <193>: 
x2sys_cross [WARNING]: No time column, use dummy times
x2sys_cross [WARNING]: No time column, use dummy times
gmtest: line 54:  3168 Segmentation fault      "$@"
Maybe you could run that script line by line and see where it actually crashes?

fitcircle/circles.sh gives a large RMS differences, what does it look like?

**Comment by PaulWessel** [comment]
How to you tell it to read the mb dll and what name do you pass? I dont this the "add _w64" is a standard thing but your own system, no?  What if you pass the full name?

**Comment by joa-quim** [comment]
> Some stuff is intriguing: Here you/we have a double C:C:

Yes. That's an example of the MinGW bash shits.

> gmtest: line 54: 3168 Segmentation fault "$@"

I didn't have any segmenatation fault when running the tests (it always pops up a `crashing windows`)

> fitcircle/circles.sh gives a large RMS differences, what does it look like?

It doesn't print the circles. And looking at the failing dir, there are no ``gcircle.txt`` or ``scircle.txt`` files. So again I suspect of a script shit.

**Comment by PaulWessel** [comment]
Remind me again why MinGW versus Ubuntu bash for Win 10, given it seems so incapable of being Unix-like?

**Comment by joa-quim** [comment]
Because Wibuntu would be testing it's own (gcc) GMT build. Tried once to make it run the pure (VS) win binaries but had a bunch of troubles (haven't tried again)

**Comment by joa-quim** [comment]
In ``CMakeList.txt`` we have
```
# If a renaming of the gmtpslpsl dll has been set in ConfigUser.cmake
if (WIN32 AND PSL_DLL_RENAME)
	set_target_properties (pslib PROPERTIES RUNTIME_OUTPUT_NAME ${PSL_DLL_RENAME})
endif (WIN32 AND PSL_DLL_RENAME)
```

and in my ``ConfigUser.cmake`` I do
```
	set (GMT_DLL_RENAME gmt_w${BITAGE})
	set (PSL_DLL_RENAME postscriptlight_w${BITAGE})
	set (SUPP_DLL_RENAME supplements_w${BITAGE})
```
I'm not worrying much about this. Must be that my ``build`` dir must be contaminated with previous build files.

**Comment by seisman** [comment]
Fixed by #943.
